### ISSUE TITLE
Cannot use `match case` to destructuring any object using `case object(x=y)`

### ISSUE THREAD

**Issue by RenDelaCruz** [python/mypy#16912]
**Bug Report**

In the global scope, mypy allows destructuring any object in `match case` with `object(x=y)`:

```py
@dataclass
class A:
    id: int

# SUCCESS
match A(id=5):
    case object(id=object_id):
        print(object_id)
```

**However**, in a method, mypy errors out trying to use `object(x=y)` in `match case`:

```py
# FAILURE
def get_id_from_any_object(obj: object) -> int | None:
    match obj:
        # main.py:19: error: Class "builtins.object" has no attribute "id"
        case object(id=object_id):
            return object_id
        case _:
            return None
```

**To Reproduce**

mypy playground:
https://mypy-play.net/?mypy=latest&python=3.12&gist=c5799b1b128c52894f2879f9fc8d0115

**Expected Behavior**

Should allow using `case object(x=y)` in `match case` syntax.

**Actual Behavior**

Errors out trying `case object(x=y)` in `match case` syntax, inside a method. Does not error out in global scope.

**Your Environment**

- Mypy version used: 1.8.0
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.12

**Comment by Hnasar** [comment]
I poked at it a little and there is a bug here but I wanted to clarify that this isn't influenced by the scope: it has to do with the type of the `match` ["subject"](https://docs.python.org/3/whatsnew/3.10.html#syntax-and-operations).

In your global scope example the subject has type `A` but in the function, the subject, `obj` has type `object`.

If we remove the match statement:
* Normally at runtime if you do `python3 -c 'o = object(); o.id'` you'll get an `AttributeError`. ✅
* Mypy also complains in this case: `python3 -m mypy -c 'o = object(); o.id'`:  ✅  
`error: "object" has no attribute "id"  [attr-defined]` 

However, with a match statment it's a bit different 

```py
match object():
    case object(id=object_id):
        print("match")
    case _:
        print("no match")
```

* No error at runtime: ✅
* Unfortunately mypy errors:  🐞🐞🐞  
`error: Class "builtins.object" has no attribute "id" `

**Comment by TeamSpen210** [comment]
The `match` statement is more like a `hasattr()` check, which mypy does have the ability to handle. But changing the behaviour to do that then opens up false negatives where mypy fails to warn you about typos etc.

**Comment by erictraut** [comment]
A similar question was recently posted to the [pyright issue tracker](https://github.com/microsoft/pyright/issues/7179). From my perspective, this is a legitimate use of pattern matching. Pyright therefore does not generate a type error [for the code at the top of this thread](https://pyright-play.net/?enableExperimentalFeatures=true&code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sBYAKHoAF4k1MN62soBBALnpQhBGHwIA7BPXoREyABa8AFPhgBeAKwBKAXWFRkmbFDAAjAFbZkCFerOXrAfVU7B%2B4URD5JS%2B1YTOMFrSdDDYwFAA5tgBqo6gkI5w4gCejn7WvhZiGQhaUAC0AHwS5AA%2BUAByYOLYuvqyCAom2W7uBkYtDjaqarmBrnrt7iAxAK4g4l3%2BgW3uhjhQjvXD%2BqMIE1PVtfRAA). As @TeamSpen210 notes, this comes at the expense of some potential false negatives.

**Comment by RenDelaCruz** [comment]
I agree, using `object()` in this case is a legitimate use of pattern matching. Perhaps mypy should make an exception for subjects with an `object` type, when used in match case statements only

**Comment by cdce8p** [comment]
Fixed in https://github.com/python/mypy/pull/19709. Will be included in the next release.
Note that I only exempted `builtins.object` so we can still detect missing attributes on other classes.
